File: python/dfs/trie.py
# Trie is a tree data structure that is used to store strings. It is used to store a dictionary of words.
import unittest
from collections import Counter


class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_of_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()
            current = current.children[char]
        current.end_of_word = True

    def search(self, word, char_counts):
        current = self.root
        for char in word:
            if char not in current.children:
                return False
            if char_counts[char] == 0:
                return False
            char_counts[char] -= 1
            current = current.children[char]
        return current.end_of_word


def longest_matching_word(dictionary, chars):
    trie = Trie()
    char_counts = Counter(chars)

    for word in dictionary:
        trie.insert(word)

    result = ""
    for word in dictionary:
        count = char_counts.copy()
        if trie.search(word, count):
            if len(word) > len(result):
                result = word
    return result
        

print(longest_matching_word(['when', 'what', 'whatthen', 'whatnow'], "whatno"))
print(longest_matching_word(['when', 'what', 'whatthen', 'whatnow'], "whatnwo"))
print(longest_matching_word(['when', 'what', 'whatthen', 'whatnow'], "wontthaw"))

File: go/dfs/trie.go
package dfs

type TrieNode struct {
	children map[rune]*TrieNode
	isWord   bool
}

type Trie struct {
	root *TrieNode
}

func constructor() *Trie {
	return &Trie{root: &TrieNode{children: make(map[rune]*TrieNode), isWord: false}}
}

func (t *Trie) insert(word string) {
	node := t.root
	for _, r := range word {
		if _, ok := node.children[r]; !ok {
			node.children[r] = &TrieNode{children: make(map[rune]*TrieNode)}
		}
		node = node.children[r]
	}
	node.isWord = true
}

func (t *Trie) search(word string, charCounts map[rune]int) bool {
	node := t.root
	for _, r := range word {
		if _, ok := node.children[r]; !ok {
			return false
		}
		if charCounts[r] == 0 {
			return false
		}
		charCounts[r]--
		node = node.children[r]
	}
	return node.isWord
}

func LongestMatchingWord(dictionary []string, chars string) string {
	trie := constructor()
	charCounts := make(map[rune]int)
	for _, r := range chars {
		charCounts[r]++
	}

	for _, word := range dictionary {
		trie.insert(word)
	}

	result := ""
	for _, word := range dictionary {
		count := make(map[rune]int)
		for k, v := range charCounts {
			count[k] = v
		}
		if trie.search(word, count) && len(word) > len(result) {
			result = word
		}
	}
	return result
}
